data-structures/multiblock-collections.md

```diff
@@ -0,0 +1,264 @@
+* **Useful** in the sense that rich applications are only practical when a basic set of data organization tools are available: maps, sets, arrays. And functionality to interact with that data is straightforward: lookups, searches, queries, mutations.
+
+IPLD, by design, removes the distinction between local and remote data sources (i.e. memory, disk, database/network). Data collection abstractions should remain consistent with this, essentially providing application programmers with simple access to potentially massive distributed data stores. With such interfaces, apparently simple data fetch operations, such as `Map->Get(Key)`, could traverse extremely large and distributed data stores without the aid of an intermediate, such as a DBMS (or perhaps without the _apparent_ aid of an intermediate since the IPFS network and its protocols provides such a service).
+
+## Immutability
+
+As a content-addressed storage system, all collections implemented on top of IPLD are immutable. That is, any particular instance of a collection represents a structure that cannot change as the content, including its links, are fixed by their addresses.
+
+It is likely that all IPLD collections will be represented by a single "root" block that provides a gateway to the entire collection—the beginning of a traversal algorithm specific to that collection (although we should not preclude the possibility of novel collections that aren't represented by a single root block). Any mutation of an IPLD collection necessarily means the alteration of content within the data structure. At a minimum this will require at least the creation of a new root block to contain such changes. For large data sets that span many blocks, mutations are likely to require a cascading replacement of many blocks. This will lead to copy-on-write semantics, where mutations of an existing data structure result in an overlay of only those blocks that must change to reflect those updates. In practical terms, many root blocks representing different data sets may share intermediate blocks if they are derived from similar sources via mutation operations; and as a content-addressed storage system it's entirely possible for collections created completely independently to share components if their content and links resolve to the same addresses.
+
+For the purpose of mutation, IPLD collection interfaces will either have to track replaced root blocks or expose copy-on-write directly by returning new root blocks when users call mutation operations.
+
+For example:
+
+```
+A = Collection(OldRoot)
+A.Add(Element1)
+A.Delete(Element2)
+NewRoot = A.Root
+```
+
+or
+
+```
+NewRoot1 = Collection(OldRoot).Add(Element1)
+NewRoot2 = Collection(OldRoot).Delete(Element2)
```

Comment:
For this example to b equivalent to the prior example I think the second line needs to operate on `NewRoot1` instead of `OldRoot`.